File: src/PYB11/ArtificialViscosity/PythonScalarArtificialViscosity.py
#-------------------------------------------------------------------------------
# PythonScalarArtificialViscosity
#-------------------------------------------------------------------------------
from PYB11Generator import *
from ArtificialViscosity import *

@PYB11template("Dimension")
@PYB11module("SpheralCompiledModules.SpheralArtificialViscosity")
class PythonScalarArtificialViscosity(ArtificialViscosity):
    """Python-overridable artificial viscosity (Scalar version, CPU-only).

    Allows rapid prototyping of new viscosity models in Python.

    WARNING: Significantly slower than C++ implementations (100-1000x slower).
             Use for prototyping only, not production runs.
             Forces serial execution (no OpenMP) due to Python GIL.

    Example usage:
        from Spheral1d import *

        class MyViscosity(PythonScalarArtificialViscosity1d):
            def __init__(self, Cl, Cq, kernel):
                PythonScalarArtificialViscosity1d.__init__(self, Cl, Cq, kernel)

            def computeQPiij(self, QPiij, QPiji, Qij, Qji,
                           xi, vi, rhoi, csi,
                           xj, vj, rhoj, csj,
                           etai, etaj,
                           fCli, fCqi, fClj, fCqj):
                # Your custom viscosity implementation here
                vij = vi - vj
                xij = xi - xj
                if vij.dot(xij) < 0.0:  # Compression only
                    mui = vij.dot(etai) / (etai.magnitude2() + self.epsilon2())
                    # ... compute QPiij, QPiji, Qij, Qji
                else:
                    QPiij = 0.0
                    QPiji = 0.0
                    Qij = 0.0
                    Qji = 0.0

            def label(self):
                return "MyViscosity"

        # Use it
        av = MyViscosity(1.0, 2.0, kernel)
        hydro = SPH1d(dataBase, av, kernel)
    """

    PYB11typedefs = """
    using Scalar = typename %(Dimension)s::Scalar;
    using Vector = typename %(Dimension)s::Vector;
    using Tensor = typename %(Dimension)s::Tensor;
    using SymTensor = typename %(Dimension)s::SymTensor;
    using ThirdRankTensor = typename %(Dimension)s::ThirdRankTensor;
    using TimeStepType = typename Physics<%(Dimension)s>::TimeStepType;
    using ResidualType = typename Physics<%(Dimension)s>::ResidualType;
    using VolumeRequirements = typename Physics<%(Dimension)s>::VolumeRequirements;
    using RKRequirements = typename Physics<%(Dimension)s>::RKRequirements;
    using ConnectivityRequirements = typename Physics<%(Dimension)s>::ConnectivityRequirements;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               Clinear = "const Scalar",
               Cquadratic = "const Scalar",
               kernel = "const TableKernel<%(Dimension)s>&"):
        "PythonScalarArtificialViscosity constructor"

    #...........................................................................
    # Methods
    @PYB11pure_virtual
    @PYB11const
    def computeQPiij(self,
                     # Particle i state
                     xi    = "const Vector&",
                     vi    = "const Vector&",
                     rhoi  = "const Scalar",
                     csi   = "const Scalar",
                     # Particle j state
                     xj    = "const Vector&",
                     vj    = "const Vector&",
                     rhoj  = "const Scalar",
                     csj   = "const Scalar",
                     # Pre-computed H*dx (eta)
                     etai  = "const Vector&",
                     etaj  = "const Vector&",
                     # Viscosity coefficient multipliers
                     fCli   = "const Scalar",
                     fCqi   = "const Scalar",
                     fClj   = "const Scalar",
                     fCqj   = "const Scalar"):
        """Compute artificial viscosity for particle pair (i,j).

        Override this method in Python to implement custom viscosity models.

        Should return results a tuple: (QPiij, QPiji, Qij, Qji)

        This is a SIMPLIFIED interface with only 14 arguments (vs 20+ in full QPiij).
        All FieldList lookups and complex data extraction is handled by C++.

        Returns a tuple:
            QPiij, QPiji: Q/rho^2 viscous pressure outputs (modify in place)
            Qij, Qji: Viscous pressure Q outputs (modify in place)

        Args:
            xi, vi, rhoi, csi: Position, velocity, density, sound speed for particle i
            xj, vj, rhoj, csj: Position, velocity, density, sound speed for particle j
            etai, etaj: Pre-computed Hi*(xi-xj) and Hj*(xj-xi)
                        Use these instead of recomputing from positions
            fCli, fCqi: Linear and quadratic viscosity multipliers for particle i
            fClj, fCqj: Linear and quadratic viscosity multipliers for particle j

        Typical implementation pattern:
            vij = vi - vj
            xij = xi - xj

            if vij.dot(xij) < 0.0:  # Compression
                mui = vij.dot(etai) / (etai.magnitude2() + self.epsilon2)
                muj = vij.dot(etaj) / (etaj.magnitude2() + self.epsilon2)

                Clij = 0.5 * (fCli + fClj) * self.Cl
                Cqij = 0.5 * (fCqi + fCqj) * self.Cq

                ei = -Clij * csi * min(0.0, mui) + Cqij * min(0.0, mui)**2
                ej = -Clij * csj * min(0.0, muj) + Cqij * min(0.0, muj)**2

                return (ei / rhoi, 
                        ej / rhoj,
                        rhoi * ei,
                        rhoj * ej)
            else:
                return (0.0, 0.0, 0.0, 0.0)
        """
        return "std::tuple<Scalar, Scalar, Scalar, Scalar>"

    @PYB11virtual
    @PYB11protected
    def updateManagedPtr(self):
        "Update member data for managed pointer."
        return "void"

    @PYB11virtual
    @PYB11const
    def QPiTypeIndex(self):
        "Require ArtificialViscosities to specify the type_index of the descendant QPiType"
        return "std::type_index"

    @PYB11virtual
    @PYB11const
    def label(self):
        return "std::string"

File: src/ArtificialViscosity/PythonScalarArtificialViscosity.cc
//---------------------------------Spheral++----------------------------------//
// PythonScalarArtificialViscosity -- Implementation
//----------------------------------------------------------------------------//
#include "ArtificialViscosity/PythonScalarArtificialViscosity.hh"
#include "Utilities/SpheralMessage.hh"

#include <stdexcept>

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor
//------------------------------------------------------------------------------
template<typename Dimension>
PythonScalarArtificialViscosity<Dimension>::
PythonScalarArtificialViscosity(const Scalar Clinear,
                                const Scalar Cquadratic,
                                const TableKernel<Dimension>& kernel) :
  ArtificialViscosity<Dimension>(Clinear, Cquadratic, kernel),
  mView(nullptr) {
}

//------------------------------------------------------------------------------
// Destructor
//------------------------------------------------------------------------------
template<typename Dimension>
PythonScalarArtificialViscosity<Dimension>::
~PythonScalarArtificialViscosity() {
  if (mView) {
    mView.free();
  }
}

//------------------------------------------------------------------------------
// Get scalar view - create wrapper on demand
//------------------------------------------------------------------------------
template<typename Dimension>
chai::managed_ptr<typename PythonScalarArtificialViscosity<Dimension>::ArtViscViewScalar>
PythonScalarArtificialViscosity<Dimension>::
getScalarView() {
  if (!mView) {
    mView = chai::make_managed<PythonAVView>(this);
  }
  return chai::dynamic_pointer_cast<ArtViscViewScalar>(mView);
}

//------------------------------------------------------------------------------
// PythonAVView Constructor
//------------------------------------------------------------------------------
template<typename Dimension>
PythonScalarArtificialViscosity<Dimension>::PythonAVView::
PythonAVView(PythonScalarArtificialViscosity<Dimension>* parent) :
  ArtificialViscosityView<Dimension, Scalar>(parent->Cl(),
                                             parent->Cq(),
                                             parent->balsaraShearCorrection(),
                                             parent->epsilon2(),
                                             parent->negligibleSoundSpeed()),
  mParent(parent) {
  REQUIRE(parent != nullptr);
}

//------------------------------------------------------------------------------
// QPiij implementation - extract FieldListView data and call Python method
//------------------------------------------------------------------------------
template<typename Dimension>
void
PythonScalarArtificialViscosity<Dimension>::PythonAVView::
QPiij(Scalar& QPiij, Scalar& QPiji,
      Scalar& Qij, Scalar& Qji,
      const size_t nodeListi, const size_t i,
      const size_t nodeListj, const size_t j,
      const Vector& xi,
      const SymTensor& Hi,
      const Vector& etai,
      const Vector& vi,
      const Scalar rhoi,
      const Scalar csi,
      const Vector& xj,
      const SymTensor& Hj,
      const Vector& etaj,
      const Vector& vj,
      const Scalar rhoj,
      const Scalar csj,
      const FieldListView<Dimension, Scalar>& fCl,
      const FieldListView<Dimension, Scalar>& fCq,
      const FieldListView<Dimension, Tensor>& DvDx) const {

  // Extract multipliers from FieldListView
  // If FieldList is empty, use default multiplier of 1.0
  const auto fCli = (fCl.size() > 0) ? fCl(nodeListi, i) : 1.0;
  const auto fCqi = (fCq.size() > 0) ? fCq(nodeListi, i) : 1.0;
  const auto fClj = (fCl.size() > 0) ? fCl(nodeListj, j) : 1.0;
  const auto fCqj = (fCq.size() > 0) ? fCq(nodeListj, j) : 1.0;

  // Call simplified Python-overridable method with exception handling
  // This catches any Python exceptions and converts them to C++ errors
  try {
    mParent->computeQPiij(QPiij, QPiji, Qij, Qji,
                          xi, vi, rhoi, csi,
                          xj, vj, rhoj, csj,
                          etai, etaj,
                          fCli, fCqi, fClj, fCqj);
  } catch (const std::exception& e) {
    // Python raised an exception - report it clearly
    VERIFY2(false, "Python computeQPiij raised exception: " << e.what());
  } catch (...) {
    // Unknown exception from Python
    VERIFY2(false, "Python computeQPiij raised unknown exception");
  }
}

}
